### ISSUE TITLE
Bitpay Stuck On Please Wait

### ISSUE THREAD

**Issue by aeridus** [spesmilo/electrum#5337]
I can't seem to get any payments to go through with Bitpay on Electrum 3.3.5. It gets stuck at "please wait..." indefinitely (>20 minutes) when I paste the payment url. However, I can still use Electrum 3.3.4 to send bitcoins to Bitpay.

**Comment by SomberNight** [comment]
Hmm... I've just paid two bitpay invoices on testnet without issues.

**Comment by aeridus** [comment]
I'm on Windows 7 64-bit if that affects anything. I also noticed the user agent string changed in 3.3.5, could that be a culprit? I've heard Bitpay blocks unsupported wallets.

**Comment by SomberNight** [comment]
> I also noticed the user agent string changed in 3.3.5, could that be a culprit?

Do you mean this from the release notes?
https://github.com/spesmilo/electrum/blob/dd7b356fcc70b774e63a407165a089178a5e506f/RELEASE-NOTES#L34-L36
No, that's unrelated. It's specific to the Electrum client-server protocol.

**Comment by moerto** [comment]
I have the same issue, bit pay wouldn't work with 3.3.6. and 3.3.5.but 3.3.4. works flawlessly.

**Comment by SomberNight** [comment]
Please [enable debug logging](https://github.com/spesmilo/electrum-docs/blob/master/faq.rst#how-to-enable-debug-logging) and see if there is anything interesting there.

**Comment by laggingreflex** [comment]
This is happening to me on v3.3.8 (Win10).

Here's my logs from `%AppData%\Electrum\logs\electrum_log_nnn.log`

Some of the errors seem to be even before I initiated the "Pay In Wallet" link on BitPay. 

```
20190814T200602.895939Z |     INFO | network | couldn't launch iface btc.cihar.com:50002:s -- CancelledError()
20190814T200603.279936Z |     INFO | interface.[hsmiths4fyqlw5xw.onion:50002] | disconnecting due to: ErrorGettingSSLCertFromServer(ConnectError(gaierror(11001, 'getaddrinfo failed'),),)
20190814T200603.280937Z |     INFO | network | couldn't launch iface hsmiths4fyqlw5xw.onion:50002:s -- CancelledError()
20190814T200614.889510Z |     INFO | network | network: retrying connections
20190814T200617.018301Z |     INFO | network | fee_histogram [[65, 115744], [53, 106558], [40, 260029], [39, 8799], [26, 128921], [2, 272219], [1, 744010]]
20190814T200617.018301Z |     INFO | network | fee_estimates {25: 1115, 10: 24471, 5: 39955, 2: 61845}
20190814T200622.665133Z |     INFO | interface.[btc.usebsv.com:50006] | connection established. version: ['ElectrumX 1.12', '1.4']
20190814T200622.816129Z |     INFO | interface.[btc.usebsv.com:50006] | set blockchain with height 590125
20190814T200622.817133Z |     INFO | interface.[btc.usebsv.com:50006] | skipping header 590125
20190814T200633.451858Z |     INFO | network | couldn't launch iface yuio.top:50002:s -- TimeoutError()
20190814T200637.026975Z |     INFO | interface.[electrum.leblancnet.us:50002] | disconnecting due to: ConnectError(ConnectionRefusedError(10061, "Connect call failed ('47.7.62.8', 50002)"),)
20190814T200637.028976Z |     INFO | network | couldn't launch iface electrum.leblancnet.us:50002:s -- CancelledError()
20190814T200658.945181Z |     INFO | simple_config.SimpleConfig | electrum directory C:\Users\xxx\AppData\Roaming\Electrum
20190814T200701.032351Z |     INFO | interface.[btc.xskyx.net:50002] | disconnecting due to: ConnectError(OSError("Multiple exceptions: [Errno 10060] Connect call failed ('2a00:1158:0:100::14', 50002, 0, 0), [Errno 10061] Connect call failed ('80.67.16.8', 50002)",),)
20190814T200701.034351Z |     INFO | network | couldn't launch iface btc.xskyx.net:50002:s -- CancelledError()
20190814T200714.929722Z |     INFO | network | network: retrying connections
20190814T200717.068740Z |     INFO | network | fee_histogram [[30, 110904], [1, 720401]]
20190814T200717.068740Z |     INFO | network | fee_estimates {25: 1115, 10: 24471, 5: 39955, 2: 61845}
20190814T200722.359859Z |     INFO | interface.[elx01.knas.systems:50002] | disconnecting due to: ConnectError(TimeoutError(10060, "Connect call failed ('192.168.100.100', 50002)"),)
20190814T200722.360858Z |     INFO | network | couldn't launch iface elx01.knas.systems:50002:s -- CancelledError()
20190814T200726.043263Z |     INFO | interface.[electrum.leblancnet.us:50002] | disconnecting due to: ConnectError(ConnectionRefusedError(10061, "Connect call failed ('47.7.62.8', 50002)"),)
20190814T200726.046283Z |     INFO | network | couldn't launch iface electrum.leblancnet.us:50002:s -- CancelledError()
20190814T200726.986201Z |     INFO | interface.[pink-gold.westeurope.cloudapp.azure.com:56002] | connection established. version: ['ElectrumX 1.10.0', '1.4']
20190814T200727.142201Z |     INFO | interface.[pink-gold.westeurope.cloudapp.azure.com:56002] | set blockchain with height 590125
20190814T200727.143202Z |     INFO | interface.[pink-gold.westeurope.cloudapp.azure.com:56002] | skipping header 590125
20190814T200727.613337Z |     INFO | interface.[yuio.top:50002] | disconnecting due to: ErrorGettingSSLCertFromServer(ConnectError(OSError("Multiple exceptions: [Errno 10060] Connect call failed ('2606:4700:30::681f:4106', 50002, 0, 0), [Errno 10060] Connect call failed ('2606:4700:30::681f:4006', 50002, 0, 0), [Errno 10060] Connect call failed ('104.31.64.6', 50002), [Errno 10060] Connect call failed ('104.31.65.6', 50002)",),),)
20190814T200729.477928Z |    ERROR | util | Exception in get_payment_request: TimeoutError()
Traceback (most recent call last):
  File "electrum\util.py", line 934, in wrapper
  File "electrum\util.py", line 808, in get_payment_request
  File "electrum\paymentrequest.py", line 83, in get_payment_request
  File "site-packages\aiohttp\client.py", line 1005, in __aenter__
  File "site-packages\aiohttp\client.py", line 575, in _request
  File "site-packages\aiohttp\helpers.py", line 585, in __exit__
concurrent.futures._base.TimeoutError
20190814T200745.874662Z |     INFO | exchange_rate.CoinDesk | getting fx quotes for USD
20190814T200746.624390Z |     INFO | exchange_rate.CoinDesk | received fx quotes
20190814T200746.627390Z |     INFO | gui.qt.history_list.HistoryModel | refreshing... reason: fx_quotes
20190814T200746.648391Z |    DEBUG | util.profiler | Abstract_Wallet.get_full_history 0.0210
20190814T200746.649391Z |    DEBUG | util.profiler | HistoryModel.refresh 0.0220
20190814T200746.669393Z |    DEBUG | util.profiler | AddressList.update 0.0200
```

**Comment by SomberNight** [comment]
Thank you for the log @laggingreflex 

It's not really clear to me exactly what's going on, especially as others say that they can pay using 3.3.4 just not with later versions.
Maybe your issue is different than OP's.
In any case, perhaps with https://github.com/spesmilo/electrum/commit/db8e6cabb451617081261ee1aefe06ab9a3f1fad at least the dialog will not be left there indefinitely and the user will get some feedback.

**Comment by aeridus** [comment]
I'm seeing "No FX rate available" despite having the same preferences set in both the Windows 7 and Windows 7 machines. I've included a full log (with file paths redacted).
[electrum_log_20190830T014115Z_15872.log](https://github.com/spesmilo/electrum/files/3557921/electrum_log_20190830T014115Z_15872.log)

**Comment by ecdsa** [comment]
your log shows the same TimeoutError as @laggingreflex 
that is fixed by db8e6ca
the "no fx rate" issue is unrelated

**Comment by SomberNight** [comment]
This now came up [again on reddit](https://www.reddit.com/r/Electrum/comments/eictvy/electrum_338_issues_please_wait_to_send_can_i/).

I think it is related to https://github.com/spesmilo/electrum/commit/9b0773cf2bf139eafe4a828c8964461d34ea9e23
We are doing the IPv6 and then the IPv4 DNS resolutions in series, compared to just doing IPv4 previously.
I've tested on a Windows machine where it takes ~8 seconds to load the bip70 payment request with IPv4 only dns (without this commit) and ~13 seconds with IPv6-then-IPv4-in-series (on master).

The timeout the users here are hitting is 30 seconds. Apparently they do not hit it with IPv4 only but do with IPv6-then-IPv4-in-series.

-----

It's tempting to just kill all the https://github.com/spesmilo/electrum/issues/4421 -related hacks we have, so I've tried that now but then my test machine too reached the 30 second timeout. That is, with native dns, without our custom hacks, I could not load the payment request either at all.

-----

So, first of all, as this is about generic networking (dns...), I think we should increase the default 30 second timeout on `make_aiohttp_session`.
Also, maybe we should make the IPv6 and IPv4 dns requests in parallel, although our hacks are getting more and more complex :(
Also, I will check whether our custom hack is actually utilising any kind of dns cache as I now suspect it does not.
